utils/AnalysisUtils.py

```diff
@@ -447,53 +301,96 @@ def GetExpectedBkgFromSideBands(hMassData, bkgFunc='pol2', nSigmaForSB=4, hMassS
+    fit = hMassData.Fit(funcBkg, 'Q+')
+    expBkg3s, errExpBkg3s = 0., 0.
+    if int(fit) == 0:
+        expBkg3s = funcBkg.Integral(mean - 3 * sigma, mean + 3 * sigma) / hMassData.GetBinWidth(1)
+        errExpBkg3s = funcBkg.IntegralError(mean - 3 * sigma, mean + 3 * sigma) / hMassData.GetBinWidth(1)
+    return expBkg3s, errExpBkg3s, hMassData
 
+def GetExpectedBkgFromSideBandsImp(hMassData, bkgFunc='pol2', nSigmaForSB=4, mean=0., sigma=0.,
+                                   meanSecPeak=0., sigmaSecPeak=0.):
+    '''
+    Helper method to get the expected bkg from side-bands, improved using maximum-likelihood and
+    background functions defined only on the sidebands
+
+    Parameters
+    ----------
+    - hMassData: invariant-mass histogram from which extract the estimated bkg
+    - bkgFunc: expression for bkg fit function
+    - nSigmaForSB: number of sigmas away from the invariant-mass peak to define SB windows
+    - mean: mean of invariant-mass peak of the signal
+    - sigma: width of invariant-mass peak of the signal
+    - meanSecPeak: mean of invariant-mass peak of the second peak (only Ds)
+    - sigmaSecPeak: width of invariant-mass peak of the second peak (only Ds)
 
-def GetExpectedBkgFromMC(hMassBkg, hMassSignal=None, mean=-1., sigma=-1., doFit=True, bkgFunc='pol3'):
+    Returns
+    ----------
+    - expBkg3s: expected background within 3 sigma from signal peak mean
+    - errExpBkg3s: error on the expected background
+    - hMassData: SB histogram with fit function (if fit occurred)
+    '''
+    numEntriesSB = hMassData.Integral(1, hMassData.FindBin(mean - nSigmaForSB * sigma))
+    numEntriesSB += hMassData.Integral(hMassData.FindBin(mean + nSigmaForSB * sigma), hMassData.GetNbinsX())
+    if meanSecPeak > 0 and sigmaSecPeak > 0:
+        numEntriesSB -= hMassData.Integral(hMassData.FindBin(meanSecPeak - nSigmaForSB * sigmaSecPeak),
+                                           hMassData.FindBin(meanSecPeak + nSigmaForSB * sigmaSecPeak))
+
+    if numEntriesSB <= 5: # check to have some entries in the histogram before fitting
+        return 0., 0., hMassData
+
+    minMass = hMassData.GetBinLowEdge(1)
+    maxMass = hMassData.GetBinLowEdge(hMassData.GetNbinsX()) + hMassData.GetBinWidth(1)
+    bkgFuncCreator = BkgFitFuncCreator(bkgFunc, minMass, maxMass, nSigmaForSB, mean, sigma, meanSecPeak, sigmaSecPeak)
+    integral = hMassData.Integral('width')
+    funcBkgSB = bkgFuncCreator.GetSideBandsFunc(integral)
+    fit = hMassData.Fit(funcBkgSB, 'LRQ+')
+    expBkg3s, errExpBkg3s = 0., 0.
+    if int(fit) == 0:
+        funcBkg = bkgFuncCreator.GetFullRangeFunc(funcBkgSB)
+        expBkg3s = funcBkg.Integral(mean - 3 * sigma, mean + 3 * sigma) / hMassData.GetBinWidth(1)
+        errExpBkg3s = funcBkg.IntegralError(mean - 3 * sigma, mean + 3 * sigma) / hMassData.GetBinWidth(1)
+    return expBkg3s, errExpBkg3s, hMassData
+
+
+def GetExpectedBkgFromMC(hMassBkg, mean=0., sigma=0., doFit=True, bkgFunc='pol3'):
```

Comment:
not necessarily in this iteration, but I would make it consistent with the "improved" version of the SB fit in case of `doFit=True`

Reply:
Ok